### PR TITLE
DesktopFileService: Rewrite using GLib AppInfoMonitor

### PR DESCRIPTION
**Pull request by davidmhewitt**
Replace all the manual directory enumeration, watching, parsing and re-parsing of desktop files with `GLib.AppInfoMonitor` and `GLib.AppInfo.get_all`.

This partially solves the issue of #440 . First-time flatpaks will now appear when you search for them, but not yet in the "browse" views. It may also help with #182 

The major benefit is the much simpler code however. The diff looks scary and looks like it changes everything, but it's mostly just changing the usage of `KeyFile` to the built-in keyfile operations on `GLib.AppInfo` and deleting cruft.

To test, search for some apps and check they appear as they did before.

### CODE REVIEW

File: src/synapse-core/desktop-file-service.vala [hunk 1]
```diff
@@ -123,33 +122,26 @@ namespace Synapse {
             return result;
         }
 
-        private void init_from_keyfile (KeyFile keyfile) {
+        private void init_from_desktop_app_info (GLib.DesktopAppInfo app_info) {
             try {
-                if (keyfile.get_string (GROUP, "Type") != "Application") {
+                if (app_info.get_string ("Type") != "Application") {
```
Comment (danrabbit) [inline src/synapse-core/desktop-file-service.vala:127]: Is that equivalent to this?
```suggestion
                if (app_info.get_string ("Type") != GLib.KeyFileDesktop.TYPE_APPLICATION) {
```

File: src/synapse-core/desktop-file-service.vala [hunk 2]
```diff
@@ -123,33 +122,26 @@ namespace Synapse {
...
                     throw new DesktopFileError.UNINTERESTING_ENTRY ("Not Application-type desktop entry");
                 }
 
-                if (keyfile.has_key (GROUP, "Categories")) {
-                    string[] categories = keyfile.get_string_list (GROUP, "Categories");
+                if (app_info.has_key ("Categories")) {
+                    string[] categories = app_info.get_string_list ("Categories");
```
Comment (danrabbit) [inline src/synapse-core/desktop-file-service.vala:132]: Same with these ones:
```suggestion
                if (app_info.has_key (GLib.KeyFileDesktop.KEY_CATEGORIES)) {
                    string[] categories = app_info.get_string_list (GLib.KeyFileDesktop.KEY_CATEGORIES);
```

File: src/synapse-core/desktop-file-service.vala [hunk 3]
```diff
@@ -158,50 +150,34 @@ namespace Synapse {
...
                 comment = app_info.get_description () ?? "";
 
                 var icon = app_info.get_icon () ??
                 new ThemedIcon ("application-default-icon");
                 icon_name = icon.to_string ();
 
-                if (keyfile.has_key (GROUP, "Terminal")) {
-                    needs_terminal = keyfile.get_boolean (GROUP, "Terminal");
+                if (app_info.has_key ("Terminal")) {
+                    needs_terminal = app_info.get_boolean ("Terminal");
```
Comment (danrabbit) [inline src/synapse-core/desktop-file-service.vala:161]: More constants:
```suggestion
                if (app_info.has_key (GLib.KeyFileDesktop.KEY_TERMINAL)) {
                    needs_terminal = app_info.get_boolean (GLib.KeyFileDesktop.KEY_TERMINAL);
```

File: src/synapse-core/desktop-file-service.vala [hunk 4]
```diff
@@ -158,50 +150,34 @@ namespace Synapse {
...
                 }
-                if (keyfile.has_key (GROUP, "OnlyShowIn")) {
-                    show_in = parse_environments (keyfile.get_string_list (GROUP, "OnlyShowIn"));
-                } else if (keyfile.has_key (GROUP, "NotShowIn")) {
-                    var not_show = parse_environments (keyfile.get_string_list (GROUP, "NotShowIn"));
+                if (app_info.has_key ("OnlyShowIn")) {
+                    show_in = parse_environments (app_info.get_string_list ("OnlyShowIn"));
+                } else if (app_info.has_key ("NotShowIn")) {
+                    var not_show = parse_environments (app_info.get_string_list ("NotShowIn"));
```
Comment (danrabbit) [inline src/synapse-core/desktop-file-service.vala:166]:
```suggestion
                if (app_info.has_key (GLib.KeyFileDesktop.KEY_ONLY_SHOW_IN)) {
                    show_in = parse_environments (app_info.get_string_list (GLib.KeyFileDesktop.KEY_ONLY_SHOW_IN));
                } else if (app_info.has_key (GLib.KeyFileDesktop.KEY_NOT_SHOW_IN)) {
                    var not_show = parse_environments (app_info.get_string_list (GLib.KeyFileDesktop.KEY_NOT_SHOW_IN));
```

File: src/synapse-core/desktop-file-service.vala [hunk 5]
```diff
@@ -158,50 +150,34 @@ namespace Synapse {
...
             } catch (Error err) {
                 string name = "Unidentified";
 
-                try {
-                    if (keyfile.has_key (GROUP, "Name")) {
-                        name = keyfile.get_string (GROUP, "Name");
-                    }
-                } catch (GLib.KeyFileError e) {}
+                if (app_info.has_key ("Name")) {
+                    name = app_info.get_string ("Name");
```
Comment (danrabbit) [inline src/synapse-core/desktop-file-service.vala:173]:
```suggestion
                if (app_info.has_key (GLib.KeyFileDesktop.KEY_NAME)) {
                    name = app_info.get_string (GLib.KeyFileDesktop.KEY_NAME);
```

**Review by danrabbit — COMMENTED** [review]
I think there's some constants from GLib.KeyFileDesktop that we can use here

**Comment by davidmhewitt** [comment]
@danrabbit I didn't know about those constants! Great suggestion, thanks :slightly_smiling_face:

File: src/synapse-core/desktop-file-service.vala [hunk 6]
```diff
@@ -252,251 +242,64 @@ namespace Synapse {
...
-                            non_hidden_desktop_files.add (dfi);
+        private void load_all_desktop_files () {
+            non_hidden_desktop_files.clear ();
+            exec_map.clear ();
+            desktop_id_map.clear ();
+
+            var app_infos = GLib.AppInfo.get_all ();
+            foreach (unowned GLib.AppInfo app in app_infos) {
+                if (app.should_show ()) {
+                    var app_info = (GLib.DesktopAppInfo)app;
```
Comment (tintou) [inline src/synapse-core/desktop-file-service.vala:279]:
```suggestion
                    unowned GLib.DesktopAppInfo app_info = (GLib.DesktopAppInfo)app;
```

**Comment by tintou** [comment]
That's a great thing, thanks for looking at the technological debt, one tiny thing and ready to go

**Review by tintou — APPROVED** [review]

**Comment by decathorpe** [comment]
Hm. I think there's small issues with this change - for example, if I install applications (I tried with Mail and Tasks), they show up immediately after installation, but with the generic "icon missing" icon in Slingshot. After rebooting, the correct icons are there.

**Comment by davidmhewitt** [comment]
I can reproduce that prior to this change so possibly not related. I'll look into it.

**Pull request merged**